common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+
+            if (trySet(ReadInProgress, Cached)) {
+                promise success value
+            } else {
+                promise failure new ConcurrentUpdateException
+            }
+        }
+
+        def trySet(expectedState: State, desiredState: State): Boolean = {
+            state.compareAndSet(expectedState, desiredState)
+        }
+    }
+
+    protected def cacheInvalidate(key: Any)(
+        implicit transid: TransactionId, logger: Logging): Unit = {
+        if (cacheEnabled) {
+            logger.info(this, s"invalidating $key")
+            cache remove key
+        }
+    }
+
+    protected def cacheLookup[Wsuper >: W](
+        key: Any,
+        future: => Future[W],
+        fromCache: Boolean = cacheEnabled)(
+            implicit ec: ExecutionContext, transid: TransactionId, logger: Logging): Future[W] = {
+
+        if (fromCache) {
+            val promise = Promise[W]
+            val desiredEntry = Entry(transid, ReadInProgress, promise.future)
+
+            //
+            // try inserting our desired entry...
+            //
+            cache(key)(desiredEntry) flatMap {
+                // ... and see what we get back
+                case actualEntry =>
+                    // logger.debug(this, s"Cache Lookup Status: the entry is currently in state ${actualEntry.state.get} ${getSize()}")
+
+                    actualEntry.state.get match {
+                        case Cached => {
+                            logger.debug(this, s"Cached read $key")
+                            makeNoteOfCacheHit(key)
+                            actualEntry.value
+                        }
+
+                        case ReadInProgress => {
+                            if (actualEntry.transid == transid) {
+                                logger.debug(this, "Read initiated");
+                                makeNoteOfCacheMiss(key)
+                                listenForReadDone(key, actualEntry, future, promise)
+                                actualEntry.value
+
+                            } else {
+                                logger.debug(this, "Coalesced read")
+                                makeNoteOfCacheHit(key)
+                                actualEntry.value
+                            }
+                        }
+
+                        case WriteInProgress | InvalidateInProgress => {
+                            logger.debug(this, "Reading around a write in progress")
+                            makeNoteOfCacheMiss(key)
+                            future
+                        }
+                    }
+            }
+
+        } else {
+            // not caching
+            future
+        }
+    }
+
+    protected def cacheUpdate(doc: W, key: Any, future: => Future[Winfo])(
+        implicit transid: TransactionId, logger: Logging, ec: ExecutionContext): Promise[Winfo] = {
+
+        val valPromise = Promise[W]
+        val promise = Promise[Winfo]
```

Comment:
I believe we do not need `val promise = Promise[Winfo]`. `listenForWriteDone` should return a `Future[Winfo]` which we could return directly instead. Removes the need of sideeffects here.